api/actions.ts

```diff
@@ -0,0 +1,282 @@
+		}, "ON DUPLICATE KEY UPDATE `date`= ?, `usercount`= ?", [date, usercount]);
+
+		if (server.id === 'showdown') {
+			await tables.userstatshistory.insert({date, usercount});
+		}
+		return {actionsuccess: true};
+	},
+	async upkeep(params) {
+		const challengeprefix = this.verifyCrossDomainRequest();
+		const res: {[k: string]: any} = {};
+		const curuser = this.user;
+		res.loggedin = curuser.loggedin;
+		let userid = '';
+		if (curuser.loggedin) {
+			res.username = curuser.name;
+			userid = curuser.id;
+		} else if (this.cookies.get('showdown_username')) {
+			res.username = this.cookies.get('showdown_username')!;
+			userid = toID(res.username);
+		}
+		if (userid !== '') {
+			const challengekeyid = !params.challengekeyid ? -1 : parseInt(params.challengekeyid, 10);
+			const challenge = params.challstr || "";
+			res.assertion = await this.session.getAssertion(
+				userid, challengekeyid, curuser, challenge, challengeprefix
+			);
+		}
+		return res;
+	},
+	async json(params) {
+		const server = this.getServer(true);
+		if (!server) {
+			throw new ActionError(`Only registered servers can make more than one request.`);
+		}
+		if (!params.json) throw new ActionError(`No JSON sent.`);
+		let json: any[];
+		try {
+			json = JSON.parse(params.json);
+		} catch (e) {
+			throw new ActionError(`Invalid JSON sent.`);
+		}
+		if (!Array.isArray(json)) {
+			throw new ActionError(`JSON sent must be an array of requests.`);
+		}
+		if (server.server !== Config.mainserver && json.length > 20) {
+			throw new ActionError(`Too many requests were sent. Send them in more batches.`);
+		}
+		const results = [];
+		for (const request of json) {
+			const dispatcher = new Dispatcher(this.request, this.response, {body: request});
+			try {
+				if (!request.act) throw new ActionError(`Must send a request type.`);
+				const data = await dispatcher.executeActions();
+				results.push(data);
+			} catch (e) {
+				if (e.name?.endsWith('ActionError')) {
+					results.push({actionerror: e.message});
+					continue;
+				}
+				throw e;
+			}
+		}
+		return results;
+	},
+	async prepreplay(params) {
+		const server = this.getServer(true);
+		let out: any = {};
+		if (!server) {
+			out.errorip = this.getIp();
+			return;
+		}
+		const extractedFormatId = `${params.id}`.match(/^([a-z0-9]+)-[0-9]+$/);
+		const formatId = `${params.format}`.match(/^([a-z0-9]+)$/);
+		if (
+			// the server must be registered
+			!server ||
+			// the server must send all the required values
+			!params.id ||
+			!params.format ||
+			!params.loghash ||
+			!params.p1 ||
+			!params.p2 ||
+			// player usernames cannot be longer than 18 characters
+			(params.p1.length > 18) ||
+			(params.p2.length > 18) ||
+			// the battle ID must be valid
+			!extractedFormatId ||
+			// the format ID must be valid
+			!formatId ||
+			// the format from the battle ID must match the format ID
+			(formatId[1] !== extractedFormatId[1])
+		) {
+			out = 0;
```

Comment:
Why assign a value to `out` right before returning (which will put `out` out of scope anyway)?

Reply:
Typo. Meant to have it return `out`.